File: Leetcode/Trapping Rain Water/trapping_rain_water.c
// Algorithm:
//      The approach is to find such points in an array, where its left and right values are 
//  greater than the current point value. Only at this case water can be stored, else it will flow out.
//  Also, the first and last point in the array cannot store water because there is no bounds.
//  So, the approach is quite simple, just find the max value till current index from left to right 
//  and from right to left. The minimum of this value subtracted with the current height gives the water
//  that can be stored at this position. Sum of water stored at all positions is the total trapped rain water.

// To find the max element of given two numbers
int max(int a , int b) {
    if(a > b)
        return a;
    return b;
}

// To find the min element of the given two numbers
int min(int a , int b) {
    if(a < b)
        return a;
    return b;
}

int trap(int* height, int heightSize){

    // If there is no elements in the array
    if(heightSize == 0)
        return 0;

    int left_max[heightSize] , right_max[heightSize];
    left_max[0] = height[0]; 
    right_max[0] = height[heightSize - 1];

    // To find the max value from left to right and from right ot left
    for(int i = 1; i < heightSize; i++) {
        left_max[i] = max(left_max[i - 1] , height[i]);
        right_max[i] = max(right_max[i - 1] , height[heightSize - i - 1]);
    }

    // The sum of water stored at each position is the final result
    int ans = 0;
    for(int i = 1; i < heightSize - 1; i++) {
        ans += min(left_max[i] , right_max[heightSize - i - 1]) - height[i];
    }
    return ans;
}

File: Leetcode/Trapping Rain Water/Trapping Rain Water.py
# https://leetcode.com/problems/trapping-rain-water/

# Approach :
# It can be seen that the amount of water above a point (X) depends on the heights of the highest bars to the left and right of it, regardless of how far apart they are.  
# The height of the water above it is equivalent to the minimum height of the highest bars around it.  
# Water can be calculated water=minimum(left_max, right_max) - elevation_X
# Consider using two pointers, one from the left and then one from the right to find the local maxima at each index.  From there we should be able to calculate the volume of water at each stage.  
# Traverse the list once, calculate the left_max for each point and save it in a list
# Travserse again to do the same with right_max.  
# Traverse once more to use this data and find the amount of water above each point.

class Solution:
     def trap(self, height: List[int]) -> int:
         # keep track of the total water as we traverse the heights
         water = 0
        
         # number of x-values on height chart
         n = len(height)
        
         # lists to store the left_max and right_max at each point in the chart
         left_max = [0] * n
         right_max = [0] * n
        
         # default values
         left_max[0] = height[0]
         right_max[n - 1] = height[n - 1]
        
         # filling in the left_max list
         for i in range(1, n):
             left_max[i] = max(left_max[i - 1], height[i])
            
         # filling in the right max list
         for i in range(n - 2, -1, -1):
             right_max[i] = max(right_max[i + 1], height[i])
            
         # calculating the amount of water
         for i in range(n):
             water += min(left_max[i], right_max[i]) - height[i]
            
         return water
